File: bench/programs/iterators/iterators.go
package main

import (
	"fmt"
	"time"
)

// Go has no lazy iterator adapters in std (pre-range-over-func idiom), so the
// filter/map/sum is a plain loop — which is how this is normally written in Go.
func run() int64 {
	var n, p int64 = 10000000, 1000000007
	var acc int64 = 0
	for i := int64(0); i < n; i++ {
		if i%3 == 0 {
			acc = (acc + (i*i)%p) % p
		}
	}
	return acc
}

func main() {
	run() // warmup
	t0 := time.Now()
	ans := run()
	dt := time.Since(t0).Nanoseconds()
	fmt.Println(ans)
	fmt.Println(dt)
}

File: bench/programs/iterators/iterators.py
import time


def run():
    n = 10000000
    p = 1000000007
    acc = 0
    # Lazy generator expression: filter + map, summed.
    for i in (x for x in range(n) if x % 3 == 0):
        acc = (acc + (i * i) % p) % p
    return acc


run()  # warmup
t0 = time.perf_counter_ns()
ans = run()
dt = time.perf_counter_ns() - t0
print(ans)
print(dt)
